evm/p2p/kademlia.py

```diff
@@ -0,0 +1,581 @@
+
+    def __ne__(self, other):
+        return not self == other
+
+    def __hash__(self):
+        return hash(self.pubkey)
+
+
+class KBucket():
+    """
+    Each k-bucket is kept sorted by time last seen—least-recently seen node at the head,
+    most-recently seen at the tail. For small values of i, the k-buckets will generally
+    be empty (as no appropriate nodes will exist). For large values of i, the lists can
+    grow up to size k, where k is a system-wide replication parameter.
+    k is chosen such that any given k nodes are very unlikely to fail within an hour of
+    each other (for example k = 20).
+    """
+    k = k_bucket_size
+
+    def __init__(self, start, end):
+        self.start = start
+        self.end = end
+        self.nodes = []
+        self.replacement_cache = []
+        self.last_updated = time.time()
+
+    @property
+    def midpoint(self):
+        return self.start + (self.end - self.start) // 2
+
+    def id_distance(self, id):
+        return self.midpoint ^ id
+
+    def nodes_by_id_distance(self, id):
+        return sorted(self.nodes, key=operator.methodcaller('id_distance', id))
+
+    def split(self):
+        """Split at the median id"""
+        splitid = self.midpoint
+        lower = KBucket(self.start, splitid)
+        upper = KBucket(splitid + 1, self.end)
+        for node in self.nodes:
+            bucket = lower if node.id <= splitid else upper
+            bucket.add(node)
+        for node in self.replacement_cache:
+            bucket = lower if node.id <= splitid else upper
+            bucket.replacement_cache.append(node)
+        return lower, upper
+
+    def remove_node(self, node):
+        if node not in self.nodes:
+            return
+        self.nodes.remove(node)
+
+    def in_range(self, node):
+        return self.start <= node.id <= self.end
+
+    @property
+    def is_full(self):
+        return len(self) == self.k
+
+    def add(self, node):
+        """Try to add the given node to this bucket.
+
+        If the node is already present, it is moved to the tail of the list, and we return None.
+
+        If the node is not already present and the bucket has fewer than k entries, it is inserted
+        at the tail of the list, and we return None.
+
+        If the bucket is full, we add the node to the bucket's replacement cache and return the
+        node at the head of the list (i.e. the least recently seen), which should be evicted if it
+        fails to respond to a ping.
+        """
+        self.last_updated = time.time()
+        if node in self.nodes:
+            self.nodes.remove(node)
+            self.nodes.append(node)
+        elif len(self) < self.k:
+            self.nodes.append(node)
+        else:
+            self.replacement_cache.append(node)
+            return self.head
+
+    @property
+    def head(self):
+        """Least recently seen"""
+        return self.nodes[0]
+
+    @property
+    def depth(self):
```

Comment:
Makes sense, although I'd rather keep it in this file rather than move to a utils module. If it turns out to be needed elsewhere later we can move it to a utils module

Reply:
For now, no big deal, but longer term, think about the following.

Generally, you'll find relatively extensive utils files in most of my repositories.  One of the reasons for this is that utils is generally strictly a private API.  When I end up with something like this that I keep in a file that contains public APIs I'll often prefix it with an underscore to denote that it isn't a public API despite being next to other public APIs.

So, if you leave it here I'd recommend adopting that naming convention (`_compute_depth` or whatever you want to name it) to ensure that there's no confusion by library users.